lightning/src/ln/outbound_payment.rs

```diff
@@ -896,6 +922,117 @@ impl OutboundPayments {
 		Ok(())
 	}
 
+	#[cfg(async_payments)]
+	pub(super) fn static_invoice_received<ES: Deref>(
+		&self, invoice: &StaticInvoice, payment_id: PaymentId, features: Bolt12InvoiceFeatures,
+		best_block_height: u32, entropy_source: ES,
+		pending_events: &Mutex<VecDeque<(events::Event, Option<EventCompletionAction>)>>
+	) -> Result<[u8; 32], Bolt12PaymentError> where ES::Target: EntropySource {
+		macro_rules! abandon_with_entry {
+			($payment: expr, $reason: expr) => {
+				$payment.get_mut().mark_abandoned($reason);
+				if let PendingOutboundPayment::Abandoned { reason, .. } = $payment.get() {
+					if $payment.get().remaining_parts() == 0 {
+						pending_events.lock().unwrap().push_back((events::Event::PaymentFailed {
+							payment_id,
+							payment_hash: None,
+							reason: *reason,
+						}, None));
+						$payment.remove();
+					}
+				}
+			}
+		}
+
+		match self.pending_outbound_payments.lock().unwrap().entry(payment_id) {
+			hash_map::Entry::Occupied(mut entry) => match entry.get() {
+				PendingOutboundPayment::AwaitingInvoice { retry_strategy, invoice_request, .. } => {
+					let invreq = invoice_request.as_ref().ok_or(Bolt12PaymentError::UnexpectedInvoice)?;
+					if !invoice.from_same_offer(invreq) {
+						return Err(Bolt12PaymentError::UnexpectedInvoice)
+					}
+					if invoice.invoice_features().requires_unknown_bits_from(&features) {
+						abandon_with_entry!(entry, PaymentFailureReason::UnknownRequiredFeatures);
+						return Err(Bolt12PaymentError::UnknownRequiredFeatures)
+					}
+					let amount_msat = match InvoiceBuilder::<DerivedSigningPubkey>::amount_msats(&invreq) {
+						Ok(amt) => amt,
+						Err(_) => {
+							debug_assert!(false, "LDK doesn't create invreqs without msat amounts");
+							abandon_with_entry!(entry, PaymentFailureReason::UnexpectedError);
+							return Err(Bolt12PaymentError::UnknownRequiredFeatures)
```

Comment:
@jkczyz wanted to call these lines to your attention. (a) I'm pretty sure the `debug_assert` is correct but want to double check, and (b) I'm not sure if it's going to need to change once we add currency support? If so, may make sense to add new error variants rather than using these ill-fitting ones. We also may want to make `InvoiceBuilder::<DerivedSigningPubkey>::amount_msats` a standalone method since it looks a bit out of place here.

Reply:
(a) We do allow an amount-less `InvoiceRequest` if `Offer::amount` is `Some(Amount::Bitcoin { .. })`, but this error case wouldn't be hit there. So the `debug_assert` message could be re-worded, IMO.

(b) Irrespective of currency support, we plan to move some amount checks to `ChannelManager` as per https://github.com/lightningdevkit/rust-lightning/pull/3175#discussion_r1729134707. But I think the scenario is different here.
- For the normal flow, if we don't set an amount in the `InvoiceRequest` for an `Offer` with a currency, then the user needs to check the amount in the invoice before paying it.
- For the static invoice flow, there is no amount in the invoice to check. Rather the user would need to supply an amount that they would like to pay.

In practical terms, the user would be given an event and would need to call methods on `ChannelManager` to given an amount, for instance. So I'd imagine any checks from `InvoiceBuilder::amount_msats` would move there instead.

Reply:
Reworded the `debug_assert` (I had been thinking of an invreq's underlying offer's amount as being part of the invreq). 

Thanks for the context on `Currency` support. I kept the rest of the code as-is since those considerations sound like they can be tackled on rebase depending on which PR lands first, but let me know if I'm missing further takeaways.